File: services/api-gateway/app/schemas/canvas.py
from datetime import datetime
from typing import Any, Optional
from pydantic import BaseModel, field_validator
import uuid


# ── Boards ────────────────────────────────────────────────────────────────────

class BoardCreate(BaseModel):
    title: str = "Новый канвас"
    description: Optional[str] = None
    template_id: Optional[str] = None


class BoardUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    viewport_x: Optional[float] = None
    viewport_y: Optional[float] = None
    viewport_zoom: Optional[float] = None
    is_public: Optional[bool] = None


class BoardOut(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    user_id: str
    title: str
    description: Optional[str] = None
    template_id: Optional[str] = None
    is_public: bool
    share_token: Optional[str] = None
    viewport_x: float
    viewport_y: float
    viewport_zoom: float
    created_at: datetime
    updated_at: datetime

    @field_validator("id", "user_id", mode="before")
    @classmethod
    def coerce_uuid(cls, v):
        return str(v)

    @field_validator("template_id", mode="before")
    @classmethod
    def coerce_uuid_opt(cls, v):
        return str(v) if v else None


# ── Widgets ───────────────────────────────────────────────────────────────────

class WidgetCreate(BaseModel):
    widget_type: str
    x: float = 100.0
    y: float = 100.0
    width: float = 300.0
    height: float = 200.0
    z_index: int = 0
    data: dict[str, Any] = {}
    style: dict[str, Any] = {}


class WidgetUpdate(BaseModel):
    x: Optional[float] = None
    y: Optional[float] = None
    width: Optional[float] = None
    height: Optional[float] = None
    z_index: Optional[int] = None
    data: Optional[dict[str, Any]] = None
    style: Optional[dict[str, Any]] = None


class WidgetBulkUpdate(BaseModel):
    """Batch update widget positions (e.g. after multi-drag)."""
    updates: list[dict[str, Any]]  # [{id, x, y, z_index?}]


class WidgetOut(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    board_id: str
    widget_type: str
    x: float
    y: float
    width: float
    height: float
    z_index: int
    data: dict[str, Any]
    style: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    @field_validator("id", "board_id", mode="before")
    @classmethod
    def coerce_uuid(cls, v):
        return str(v)


# ── Connections ───────────────────────────────────────────────────────────────

class ConnectionCreate(BaseModel):
    from_widget_id: str
    to_widget_id: str
    style: dict[str, Any] = {}
    label: Optional[str] = None


class ConnectionOut(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    board_id: str
    from_widget_id: str
    to_widget_id: str
    style: dict[str, Any]
    label: Optional[str] = None
    created_at: datetime

    @field_validator("id", "board_id", "from_widget_id", "to_widget_id", mode="before")
    @classmethod
    def coerce_uuid(cls, v):
        return str(v)


# ── Full board detail ─────────────────────────────────────────────────────────

class BoardDetailOut(BoardOut):
    widgets: list[WidgetOut] = []
    connections: list[ConnectionOut] = []


# ── Templates ─────────────────────────────────────────────────────────────────

class TemplateOut(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    name: str
    description: Optional[str] = None
    category: str
    thumbnail_emoji: str
    template_data: dict[str, Any]
    is_system: bool

    @field_validator("id", mode="before")
    @classmethod
    def coerce_uuid(cls, v):
        return str(v)


# ── Widget live data ──────────────────────────────────────────────────────────

class ProductWidgetData(BaseModel):
    external_product_id: str
    title: str
    image_url: Optional[str] = None
    price: Optional[str] = None
    stock: int = 0
    store_name: Optional[str] = None
    marketplace: Optional[str] = None


class LogisticsWidgetData(BaseModel):
    external_product_id: str
    title: str
    image_url: Optional[str] = None
    stock: int
    # Days of supply (rough estimate: stock / avg_daily_sales; placeholder)
    days_supply: Optional[int] = None
    status: str  # ok | warn | critical


class CampaignWidgetData(BaseModel):
    campaign_id: str
    name: str
    marketplace: str
    ad_platform: str
    is_active: bool
    budget: Optional[str] = None
    utm_source: Optional[str] = None


# ── AI command ────────────────────────────────────────────────────────────────

class AICommandRequest(BaseModel):
    command: str


class AICommandResponse(BaseModel):
    message: str
    actions: list[dict[str, Any]] = []

File: frontend/src/api/canvas.ts
import api from "./client";

// ── Types ─────────────────────────────────────────────────────────────────────

export type WidgetType =
  | "product_card"
  | "logistics"
  | "ad_connector"
  | "mini_chart"
  | "sticker"
  | "text"
  | "kpi_table";

export interface CanvasWidget {
  id: string;
  board_id: string;
  widget_type: WidgetType;
  x: number;
  y: number;
  width: number;
  height: number;
  z_index: number;
  data: Record<string, any>;
  style: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface CanvasConnection {
  id: string;
  board_id: string;
  from_widget_id: string;
  to_widget_id: string;
  style: Record<string, any>;
  label?: string | null;
  created_at: string;
}

export interface CanvasBoard {
  id: string;
  user_id: string;
  title: string;
  description?: string | null;
  template_id?: string | null;
  is_public: boolean;
  share_token?: string | null;
  viewport_x: number;
  viewport_y: number;
  viewport_zoom: number;
  created_at: string;
  updated_at: string;
}

export interface CanvasBoardDetail extends CanvasBoard {
  widgets: CanvasWidget[];
  connections: CanvasConnection[];
}

export interface BoardTemplate {
  id: string;
  name: string;
  description?: string | null;
  category: string;
  thumbnail_emoji: string;
  template_data: Record<string, any>;
  is_system: boolean;
}

// Widget data types
export interface ProductWidgetData {
  external_product_id: string;
  title: string;
  image_url?: string | null;
  price?: string | null;
  stock: number;
  store_name?: string | null;
  marketplace?: string | null;
}

export interface LogisticsWidgetData {
  external_product_id: string;
  title: string;
  image_url?: string | null;
  stock: number;
  days_supply?: number | null;
  status: "ok" | "warn" | "critical";
}

export interface CampaignWidgetData {
  campaign_id: string;
  name: string;
  marketplace: string;
  ad_platform: string;
  is_active: boolean;
  budget?: string | null;
  utm_source?: string | null;
}

// ── Boards ────────────────────────────────────────────────────────────────────

export const listBoards = () =>
  api.get<CanvasBoard[]>("/canvas/boards").then((r) => r.data);

export const createBoard = (payload: { title?: string; description?: string; template_id?: string }) =>
  api.post<CanvasBoardDetail>("/canvas/boards", payload).then((r) => r.data);

export const getBoard = (id: string) =>
  api.get<CanvasBoardDetail>(`/canvas/boards/${id}`).then((r) => r.data);

export const updateBoard = (
  id: string,
  payload: Partial<{
    title: string;
    description: string;
    viewport_x: number;
    viewport_y: number;
    viewport_zoom: number;
    is_public: boolean;
  }>
) => api.patch<CanvasBoard>(`/canvas/boards/${id}`, payload).then((r) => r.data);

export const deleteBoard = (id: string) => api.delete(`/canvas/boards/${id}`);

export const toggleShareBoard = (id: string) =>
  api.post<CanvasBoard>(`/canvas/boards/${id}/share`).then((r) => r.data);

// ── Widgets ───────────────────────────────────────────────────────────────────

export interface WidgetCreatePayload {
  widget_type: WidgetType;
  x?: number;
  y?: number;
  width?: number;
  height?: number;
  z_index?: number;
  data?: Record<string, any>;
  style?: Record<string, any>;
}

export const addWidget = (boardId: string, payload: WidgetCreatePayload) =>
  api.post<CanvasWidget>(`/canvas/boards/${boardId}/widgets`, payload).then((r) => r.data);

export const updateWidget = (
  boardId: string,
  widgetId: string,
  payload: Partial<{ x: number; y: number; width: number; height: number; z_index: number; data: Record<string, any>; style: Record<string, any> }>
) => api.patch<CanvasWidget>(`/canvas/boards/${boardId}/widgets/${widgetId}`, payload).then((r) => r.data);

export const bulkUpdateWidgets = (
  boardId: string,
  updates: Array<{ id: string; x?: number; y?: number; z_index?: number }>
) => api.post(`/canvas/boards/${boardId}/widgets/bulk`, { updates });

export const deleteWidget = (boardId: string, widgetId: string) =>
  api.delete(`/canvas/boards/${boardId}/widgets/${widgetId}`);

// ── Connections ───────────────────────────────────────────────────────────────

export const addConnection = (
  boardId: string,
  payload: { from_widget_id: string; to_widget_id: string; style?: Record<string, any>; label?: string }
) => api.post<CanvasConnection>(`/canvas/boards/${boardId}/connections`, payload).then((r) => r.data);

export const deleteConnection = (boardId: string, connId: string) =>
  api.delete(`/canvas/boards/${boardId}/connections/${connId}`);

// ── Templates ─────────────────────────────────────────────────────────────────

export const listTemplates = () =>
  api.get<BoardTemplate[]>("/canvas/templates").then((r) => r.data);

// ── Widget data ───────────────────────────────────────────────────────────────

export const getProductWidgetData = (store_id: string, external_product_id: string) =>
  api
    .get<ProductWidgetData>("/canvas/widget-data/product", { params: { store_id, external_product_id } })
    .then((r) => r.data);

export const getLogisticsWidgetData = (store_id: string, external_product_id: string) =>
  api
    .get<LogisticsWidgetData>("/canvas/widget-data/logistics", { params: { store_id, external_product_id } })
    .then((r) => r.data);

export const getCampaignWidgetData = (campaign_id: string) =>
  api
    .get<CampaignWidgetData>("/canvas/widget-data/campaign", { params: { campaign_id } })
    .then((r) => r.data);

// ── AI ────────────────────────────────────────────────────────────────────────

export const sendAICommand = (boardId: string, command: string) =>
  api
    .post<{ message: string; actions: any[] }>(`/canvas/boards/${boardId}/ai-command`, { command })
    .then((r) => r.data);
